macros/contexts/contextBoolean.pl

```diff
@@ -0,0 +1,253 @@
+		'F' =>
+			{ value => context::Boolean::Value->new(0), isConstant => 0, string => 'F', TeX => '\bot', perl => 0 },
+		# 0 and 1 are not recognized unless added as constants
+		# it might not be a bad idea anyhow?
+		'1'     => { alias => 'T' },
+		'0'     => { alias => 'F' },
+		'True'  => { alias => 'T' },
+		'False' => { alias => 'F' },
+	);
+
+	## Define our logic operators
+	$context->operators->are(
+		'or' => {
+			class         => 'context::Boolean::BOP::or',
+			precedence    => $orPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' or ',
+			TeX           => '\vee ',
+			perl          => '||',
+			alteratives   => ['\x{2228}'],
+		},
+		'and' => {
+			class         => 'context::Boolean::BOP::and',
+			precedence    => $andPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' and ',
+			TeX           => '\wedge ',
+			perl          => '&&',
+			alteratives   => ['\x{2229}'],
+		},
+		'xor' => {
+			class         => 'context::Boolean::BOP::xor',
+			precedence    => $andPrec,
+			associativity => 'left',
+			type          => 'bin',
+			string        => ' xor ',
+			perl          => '!=',
+			TeX           => '\oplus ',
+			alteratives   => [ '\x{22BB}', '\x{2295}' ],
+		},
+		# unary negation is only allowed if the '-' token is recognized, so make it BOP::undef
+		'-' => {
+			class         => 'Parser::BOP::undefined',
+			precedence    => $notPrec - 1,
+			associativity => 'left',
+			type          => 'both',
+			string        => '-',
+		},
+		'not' => {
+			class            => 'context::Boolean::UOP::not',
+			precedence       => $notPrec,
+			associativity    => 'left',
+			type             => 'unary',
+			string           => ' not ',
+			TeX              => '\mathord{\sim}',
+			perl             => '!',
+			alteratives      => [ '\x{172}', '\x{732}' ],
+			nofractionparens => 1,
+		},
+		' '  => { alias => 'and' },
+		'*'  => { alias => 'and' },
+		'+'  => { alias => 'or' },
+		'!'  => { alias => 'not' },
+		'u-' => { alias => 'not' },
+		'><' => { alias => 'xor' },
+	);
+
+	## Set default variables 'p' and 'q'
+	$context->variables->are(
+		p => { isValue => 1, type => Value::Type('Boolean') },
+		q => { isValue => 1, type => Value::Type('Boolean') },
+	);
+}
+
+## Subclass Value::Formula for boolean formulas
+package context::Boolean::Formula;
+our @ISA = ('Value::Formula');
+
+sub new {
+	my $class = shift;
+	my $self  = $class->SUPER::new(@_);
+
+	my @variables = $self->{context}->variables->names;
+	my @points;
+	my @values;
+
+	my $f = $self->{f} = $self->perlFunction(undef, \@variables);
+
+	foreach my $combination (0 .. 2**@variables - 1) {
+		my @point = map { $combination & 2**$_ ? 1 : 0 } (0 .. $#variables);
+		my $value = &$f(@point) ? 1 : 0;
+		push @points, \@point;
+		push @values, $value;
+	}
+
+	$self->{test_points} = \@points;
+	$self->{test_values} = \@values;
+
+	return $self;
+}
```

Comment:
Yep, I did fix this in the time since the latest push. But your point about using proper MathObjects for the point-values is well taken. (I also overloaded createPointValues, but I can see that is not necessary.)

I suppose I could also create and stash `test_values` here in `createRandomPoints` (as the base Formula class does)...

Reply:
Actually, I'm reconsidering my recommendation about using `$T` and `$F`, here.  It turns out that I misremembered how the test points are handled.  I had thought that the coordinated were converted to MathObjects when they are created, but that turns out not to be the case; they remain perl reals, and are only converted to MathObjects when they are used with other math objects within the Formula, as needed.

So it turns out to be advantageous in your case to keep these as 1 and 0 rather than Boolean objects.  The reason is that the `perl` methods use `$result` as the condition in `"($result ? 1 : 0)"`, but `$result` will be something like `$p || $q`, and when `$p` and `$q` are `Boolean` objects, they will always be truthy, and so the condition will not be properly checked.  One could use

``` perl
sub perl {
	my $self   = shift;
        my $bop    = $self->{def}{perl} || $self->{def}{string};
        my $result = $self->{lop}->perl(1) . '->value ' . $bop . ' ' . $self->{rop}->perl(2) . '->value';
	return "($result ? \$context::Boolean::T : \$context::Boolean::F)";
}
```

for the `context::Boolean::BOP` package, and a similar one for `not`, but that's a lot of unnecessary packaging and unpacking of the values.  Leaving the test points with coordinates of 1 and 0 means you can use the simpler

``` perl
sub perl {
	my $self   = shift;
        my $result = $self->SUPER::perl(@_);
	return "($result ? 1 : 0)";
}
```

If you do go back to 1 and 0, then you should add `perl => 1` to the constant `T` and `perl => 0` to the constant `F`.

I'm not sure which I think is best.  The perl numbers are more efficient, but the MathObjects don't rely on a fragile setup.  Do whichever you like best.

Reply:
Oh, if you go with MathObjects, then using `perl => '$context::Boolean::T'` for the constant `T` and `perl -> '$constant::Boolean::F'` for `F` will avoid having to create new objects every time they are used in a formula.

Reply:
The only thing that I still cannot figure out with the MathObject approach is support for generating perl code from a Formula with specific values for the variables. It always seems to stringifie the object-values for `p` and `q` in the resulting perl code.

Example.pg code:

```perl
Context("Boolean");
$T = Boolean('T');
$F = Boolean('F');
$g = Formula('-(p+q)');
TEXT($g->perl(p=>$T,q=>$F));
```
Produces
```perl
context::Boolean::Boolean->new((! (T->value || F->value ? $context::Boolean::T : $context::Boolean::F)->value ? $context::Boolean::T : $context::Boolean::F))
```

Reply:
This happens even with context constants 'T' and 'F' defined with `perl => '$context::Boolean::T' (F resp)

Reply:
It looks as though `Parser::Variable::perl` returns the raw value that has been set for the variable in the parser `equation`.

https://github.com/openwebwork/pg/blob/d2bead5b9bc643ccdb1cf43c5eb302c7f5009232/lib/Parser/Variable.pm#L125-L132

So it seems this is also broken for Values that need specific perl, e.g. `Value::Complex` or `Value::String`

Reply:
The logic here is not great, but how about replacing L128 with:
```perl
if ( defined($value) ) {
    return $value->perl if ( $value->can('perl') );
    return $value;
}
```

So long as `$value` is a MathObject, it should inherit `perl` from Value.pm, and if it's not a MathObject... well, you get what you get, I guess?

Reply:
Wow, I forgot that you could specify substitutions in `Formula->perl()`.  I suspect this is not very often used, since this error has not come up in the past 20 years.

You are right this should use `$value->perl`, but probably better as `$value->perl(@_)` so that the parameters are passed on.

Note that the `$value` will always be a MathObject, due to line

https://github.com/openwebwork/pg/blob/26f77df8ca293e6ee4dc5580217a9b2994da1063/lib/Parser.pm#L862

in the `Parser::setValues()` method that is used to create the `{equation}{values}` hash.  So you should be able to do

``` perl
 	return $value->perl(@_) if defined($value); 
```

without worry.